### ISSUE TITLE
Memory leak with InfiniteChannel

### ISSUE THREAD

**Issue by tyranron** [eapache/channels#27]
If I understand correctly, the [goroutine `ch.infiniteBuffer()`](https://github.com/eapache/channels/blob/v1.1.0/infinite_channel.go#L19) persists in memory [until all the valued from inner buffer are drained](https://github.com/eapache/channels/blob/v1.1.0/infinite_channel.go#L64-L67).

So, in situation where there are many writes, but only few reads, and then both writer/reader are done with a channel (channel goes out of their scopes), the channel won't be garbage collected as the goroutine `ch.infiniteBuffer()` still runs (and we have no runtime support for our lib as for native Go channels). Which in certain circumstances (and naive usage) may lead to memory leaks.

@eapache please, correct me if I'm wrong. If not, maybe this fact needs additional mention in docs?  
[There is](https://github.com/eapache/channels/blob/v1.1.0/channels.go#L25-L27) already related one, but it points to another caveat.

**Comment by eapache** [comment]
You are not wrong, this is another caveat. In general this is true of all of the buffered channel implementations here (RingChannel, InfiniteChannel, etc):

> If the channel passes out of scope without being explicitly closed, it will leak a goroutine plus whatever values were stored in the channel at the time it went out of scope.

How does that sound?

**Comment by eapache** [comment]
Right, good point.

**Comment by michilu** [comment]
How do you think about providing another way such as supports the cancel event of `context.Context` for return from goroutines?

**Comment by tyranron** [comment]
@michilu I can barely imagine how that can be applied in this situation. Would you be so kind to provide a sort of sketch/pseudocode to illustrate your idea?

**Comment by michilu** [comment]
The inner goroutine returns if a case of closing all input channels.

This is other use cases, I try implements the Multiple-Input/Multiple-Output now.
I want do not closing edge channels of input/output each time because keep the channels owned by the worker func.
But It does not work with the Multiplex/Tee funcs.

channels:
`edge inputs => Multiplex -> WeakTee => edge outputs`

Maybe the WeakTee works well, so it has nothing to do to edge channels of output if closed the input channel.
I must be close all input channels if stop the Multiplex/WeakMultiplex goroutine, otherwise leak goroutines of the Multiplex func.

I think to it useful to supports `context.Context` for this countermeasure.

signature:
`Multiplex(ctx context.Context, output SimpleInChannel, inputs ...SimpleOutChannel)`

```go
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
Multiplex(ctx, output, inputs)
```

**Comment by tyranron** [comment]
@michilu yup, now I see... polluting channels with `context.Context` will give an ability to cancel inner runners of such channels. This definitely should work.
